### PR TITLE
[CARBONDATA-2686] Implement Left join on MV datamap

### PR DESCRIPTION
**Pull request by ravipesala**
Code ported from another branch done by @eychih to implement Left outer join. 

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5578/

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6824/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5622/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6841/

**Comment by brijoobopanna** [comment]
retest sdv please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5640/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5630/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5639/

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/DefaultMatchMaker.scala [hunk 1]
```diff
@@ -127,8 +126,27 @@ object SelectSelectNoChildDelta extends DefaultMatchPattern with PredicateHelper
     }
   }
 
-  def apply(
-      subsumer: ModularPlan,
+  private def isLeftJoinView(subsumer: ModularPlan): Boolean = {
+    if (subsumer.isInstanceOf[modular.Select]) {
+      val sel = subsumer.asInstanceOf[modular.Select]
```
Comment (jackylk) [inline datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/DefaultMatchMaker.scala]: better to use match clause to simply the logic in this function

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/DefaultMatchMaker.scala [hunk 2]
```diff
@@ -245,15 +269,22 @@ object SelectSelectNoChildDelta extends DefaultMatchPattern with PredicateHelper
...
-                Seq(done)
+                !sel_1a.predicateList.exists(_.semanticEquals(p))
+              } ++ (if (isLeftJoinView(sel_1a) &&
+                        sel_1q.joinEdges.head.joinType == Inner) {
+                sel_1a.children(1)
+                  .asInstanceOf[HarmonizedRelation].tag.map(IsNotNull(_)).toSeq
+              } else {
+                Seq.empty
+              })
+              val wip = sel_1q.copy(
```
Comment (jackylk) [inline datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/DefaultMatchMaker.scala]: can you give a better name for this variable

File: datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/Harmonizer.scala [hunk 3]
```diff
@@ -36,17 +36,33 @@ abstract class Harmonizer(conf: SQLConf)
   def batches: Seq[Batch] = {
     Batch(
       "Data Harmonizations", fixedPoint,
-      HarmonizeDimensionTable,
-      HarmonizeFactTable) :: Nil
+      Seq( HarmonizeDimensionTable) ++
+      extendedOperatorHarmonizationRules: _*) :: Nil
+//      HarmonizeFactTable) :: Nil
```
Comment (jackylk) [inline datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/Harmonizer.scala]: If it is not require, remove it

File: datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/Harmonizer.scala [hunk 4]
```diff
@@ -36,17 +36,33 @@ abstract class Harmonizer(conf: SQLConf)
...
+object FullHarmonizer extends FullHarmonizer
+
+class FullHarmonizer extends Harmonizer(new SQLConf()) {
+  override def extendedOperatorHarmonizationRules: Seq[Rule[ModularPlan]] =
+    super.extendedOperatorHarmonizationRules ++ (HarmonizeFactTable :: Nil)
 }
 
 /**
- * An default Harmonizer
+ * A semi Harmonizer - harmonize dimension tables only
```
Comment (jackylk) [inline datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/Harmonizer.scala]: Can you add description for semi harmonizer?

File: datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/ModularRelation.scala [hunk 5]
```diff
@@ -86,10 +87,28 @@ object HarmonizedRelation {
       Select(_, _, _, _, _, dim :: Nil, NoFlags, Nil, Nil, _),
       NoFlags,
       Nil, _) if (dim.isInstanceOf[ModularRelation]) =>
-        if (g.outputList
-          .forall(col => col.isInstanceOf[AttributeReference] ||
-                         (col.isInstanceOf[Alias] &&
-                          col.asInstanceOf[Alias].child.isInstanceOf[AttributeReference]))) {
+        if (g.outputList.forall(col => {
```
Comment (jackylk) [inline datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/ModularRelation.scala]: better to move the logic out of if condition. and have it better formatted

Comment (ravipesala) [reply datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/ModularRelation.scala]: ok

File: datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/Harmonizer.scala (hunk 4 above)
Comment (ravipesala) [reply datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/Harmonizer.scala]: ok

File: datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/Harmonizer.scala (hunk 3 above)
Comment (ravipesala) [reply datamap/mv/plan/src/main/scala/org/apache/carbondata/mv/plans/modular/Harmonizer.scala]: ok

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/DefaultMatchMaker.scala (hunk 2 above)
Comment (ravipesala) [reply datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/DefaultMatchMaker.scala]: ok

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/DefaultMatchMaker.scala (hunk 1 above)
Comment (ravipesala) [reply datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/DefaultMatchMaker.scala]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5657/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6881/

**Comment by jackylk** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5676/

**Pull request closed without merging**